Add section about liquidity griefing attacks

t-bast opened the pull request:
This is a PR on the `nifty/dual-fund` branch (for https://github.com/lightning/bolts/pull/851).

The liquidity griefing attack found by @morehouse is non-trivial and worth mentioning, otherwise implementations may forget to correctly handle it which would result in an easy way of attacking nodes' on-chain liquidity, creating a large opportunity cost.

morehouse commented on 02-peer-protocol.md:
Nit: the wording is slightly confusing -- does the "don't" also apply to "actively reuse them" or not?

```suggestion
It is thus recommended that implementations keep UTXOs unlocked and actively
reuse them in concurrent sessions, which guarantees that transactions created
```

t-bast replied:
Thanks, agreed! Done in https://github.com/niftynei/lightning-rfc/pull/12/commits/4c300e597464dbf110648edec387b5f8e6e8bfaa

morehouse commented on 02-peer-protocol.md:
Perhaps we should add that UTXOs should be unlocked for the whole process, even after `tx_signatures` are exchanged?  On one hand, this could help defend against pinning attacks or cases where unconfirmed inputs are used.  But on the other hand, it could increase honest UTXO conflicts, especially when trying to save on funding fees.

CC: @ariard

t-bast replied:
I added a sentence in https://github.com/niftynei/lightning-rfc/pull/12/commits/4c300e597464dbf110648edec387b5f8e6e8bfaa to address that, let me know what you think.

ariard replied:
So I agree withholding UTXO locking until the collaborative transaction confirms harden against RBF-optout pinning attacks, as a honest collaborative transaction taker can have N "honest" pending transactions confirm, and when a UTXO spend confirms, locks it and start to build on the dual-channel/splicing session, prune it from local state the other ones.

I think I see two immediate downsides to this approach. The first one, if you're doing local fee-bumping of your own unconfirmed collaborative transaction, you need to link your UTXO locking with your fee-bumping engine, otherwise you might enter in a feerate race with yourself. Of course, you could randomly "elect" an unconfirmed collaborative transaction to single-thread the fee-bumping, though there is no guarantee your counterparty does not CPFP a non-final single-owned output, breaking your local fee-estimation expectation.

The second one, if you start to have reputation engine to score collaborative transaction takers, as we have for HTLC forwarding by routing hops, you might downgrade your reputation towards the N-1 hops from doing N negotiation sessions after `tx_signatures` release. At least, we'll introduce coupling effect between the collaborative transaction flow and future (potential) reputation algorithms here, and we might have to revisit with (maybe) expensive redeployment of this funding logic, I would say.

t-bast replied:
> I think I see two immediate downsides to this approach. The first one, if you're doing local fee-bumping of your own unconfirmed collaborative transaction, you need to link your UTXO locking with your fee-bumping engine, 

I don't understand the concern, this seems very implementation-specific, I can't see anything we did in eclair that required this...especially since fee-bumping is trivial most of the time: you reuse the same input(s) and just lower the change output.

> The second one, if you start to have reputation engine to score collaborative transaction takers, as we have for HTLC forwarding by routing hops, you might downgrade your reputation towards the N-1 hops from doing N negotiation sessions after tx_signatures release.

I think it would be a bad idea to funding failure to downgrade reputation, you should probably instead use signing latency (which will be independent of actual funding success, but indicates whether the peers was witholding signatures or not, which is what we care about).

ariard replied:
I re-read the current version `a5a3a22`, and I still think the "UTXOs should only be locked once a fully signed transaction has been successfully broadcast" might still present pinning issues.

> I don't understand the concern, this seems very implementation-specific, I can't see anything we did in eclair that >required this...especially since fee-bumping is trivial most of the time: you reuse the same input(s) and just lower the > change output.

I concede the first half of the point is very-implementation specific. In LDK, currently we don't visibility for the funding transaction spent TXOs and if we have to lock-them with our fee-bumping engine, we might have to rework our interface boundary with client code.

That said, I think the second half of the point still hold, your counterparty can knock-out you from CPFP fee-bumping by chaining junk branch of transactions on announced outputs and nullifies the mempool carve-out exemption, iirc ? So from my understanding, if you're locking UTXOs at broadcast time you can still be prevented to fee-bump the dual-funding transaction ?

> I think it would be a bad idea to funding failure to downgrade reputation, you should probably instead use signing >latency (which will be independent of actual funding success, but indicates whether the peers was witholding >signatures or not, which is what we care about).

I think you would like to score our collaborative transaction taker fee-bumping strategy as otherwise they're wasting the timevalue of your contributed inputs, in addition of signing latency for sure. Though here, I think we can solve this issue by introducing a flag `concurrent_sessions_until_confirmation=true` or `concurrent_sessions_until_signature=true` in the future, to be consumed by reputation engine and let them used those flags as they wish :)

morehouse commented on 02-peer-protocol.md:
I'm not a pinning expert, but I think locking after broadcast leaves us open to the following griefing attack.

1.  Mallory constructs a funding tx with Alice using Mallory's `utxo_a`.
2.  At the same time that Mallory sends her `tx_signatures`, she also widely broadcasts a conflicting tx spending `utxo_a`.  The conflicting tx is minimum-feerate and RBF-disabled to avoid mempool replacement.  Perhaps Mallory sends the tx directly to mining pools to increase the attack success rate.
3. Alice successfully broadcasts the funding tx to her peers, but fails to replace the tx in miners' mempools.
4. Alice's funds are now locked until Mallory's conflicting tx confirms.

This attack is less trivial to pull off than the original one, and I think `mempoolfullrbf` will provide some defense if it becomes default.  But maybe there's something we can do to avoid this?

One idea is to keep UTXOs unlocked until the funding tx is *confirmed*.   This would avoid the liquidity lockup at the cost of increased chance of double-spending against honest peers.

ariard replied:
Yes I think this pinning holds and the damage provoked is a timevalue Dos, no new knowledge to me. This is correct `mempoolfullrbf` should harden the attack, however this is still triggerable even assuming full-deployment of `mempoolfullrbf` by regular nodes and miners mempools as you can partition mempools on policy differences (e.g using a non-standard `OP_RETURN` payload sent out-of-band to few "known" miners).

There has been discussion of a new p2p extension to relay conflicts at the base-layer, and address this issue: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-November/021142.html (however still at the whiteboard step).

On the idea to keep UTXO unlocked until the funding tx is confirmed, I think my comments above expressed few trade-offs to weigh in.

t-bast replied:
> One idea is to keep UTXOs unlocked until the funding tx is confirmed. This would avoid the liquidity lockup at the cost of increased chance of double-spending against honest peers.

This **vastly** increases the probability of failures with honest nodes, so I'm not sure the trade-off is worth it. The attacker needs to eclipse your bitcoin node to pull off this attack, which is a fundamentally more serious attack. If the attacker is able to eclipse you, they have other, more interesting attacks they can pull off that steal your funds, so they probably wouldn't care about griefing your liquidity.

I agree we could mention this somehow in the paragraph, but I don't think most implementations will choose to keep utxos locked until confirmation. Do you have a suggestion for how to express that?

morehouse replied:
> This **vastly** increases the probability of failures with honest nodes, so I'm not sure the trade-off is worth it. The attacker needs to eclipse your bitcoin node to pull off this attack, which is a fundamentally more serious attack.

I don't think eclipsing is a strict requirement, though it would give a 100% probability of successfully pulling off the attack.

> I agree we could mention this somehow in the paragraph, but I don't think most implementations will choose to keep utxos locked until confirmation. Do you have a suggestion for how to express that?

If we don't have a better solution today, maybe it's best to leave it as is.  Too much discussion of tradeoffs here may confuse the recommendation.

Perhaps this is an issue to address in future iterations of the protocol.

t-bast replied:
> I don't think eclipsing is a strict requirement, though it would give a 100% probability of successfully pulling off the attack.

You're right, eclipse attack isn't necessary if the honest node cannot RBF the transaction. But I expect participants to actively RBF transactions that don't confirm, and if your peer rejects your RBF attempts, it's a good indication that they're malicious...

> If we don't have a better solution today, maybe it's best to leave it as is. Too much discussion of tradeoffs here may confuse the recommendation.
> Perhaps this is an issue to address in future iterations of the protocol.

Agreed, we know that future iterations will need both sides to prevent some kind of proof-of-ownership that should protect against DoS concerns more generally. It's far from trivial to get right, but we'll have to do this at some point!

ariard replied:
> You're right, eclipse attack isn't necessary if the honest node cannot RBF the transaction. But I expect participants to actively RBF transactions that don't confirm, and if your peer rejects your RBF attempts, it's a good indication that they're malicious...

Well without going to pulling off an eclipse attack, I think you can easily pin low-feerate dual-funding transactions by abusing the mempool package limits, and as we don't check added outputs for the OP_1 CSV encumbrance, a counterparty can easily prevent you to attach new CPFP, I think.

I buy Matt suggestion to keep recommendation simple for now and iterate when we have new tools (e.g package relay/nversion=3).

morehouse reviewed (approved):
One final nit:  s/utxo/UTXO.

niftynei commented:
ACK. this is inline with previous discussions we've had regarding the reservation of UTXOs.

@t-bast can you squash the fixups and i'll merge it?

ariard reviewed (approved):
ACK a5a3a22-modulo "UTXOs should only be locked once a fully signed transaction has been successfully broadcast" can be unsafe in face of multiple flavours of pinnings (mempool package limits exhaustions, RBF rule 3, RBF opt-out double-spends).

That's okay, package-relay/nversion=3 should improve the pinning issues and all of that are recommendations only.

t-bast commented:
Thanks for the feedback, I squashed everything and removed "UTXOs should only be locked once a fully signed transaction has been successfully broadcast" in https://github.com/niftynei/lightning-rfc/pull/12/commits/60d4e84e01531ba70a0f066db740c50b835dc3e3

This way there is no recommendation on when (if ever) to lock utxos, just a recommendation to generally avoid locking them. Implementations can then choose whether they want to lock after broadcasting the transaction (which is bitcoind's behavior) or be more conservative than that to protect against pinning.

morehouse reviewed (approved):
ACK 60d4e84e01531ba70a0f066db740c50b835dc3e3
